--- linux-3.6.3/drivers/net/wireless/rtlwifi/Makefile	2012-10-21 14:32:56.000000000 -0200
+++ linux-3.6.3-devel/drivers/net/wireless/rtlwifi/Makefile	2013-06-21 10:34:39.457298356 -0300
@@ -7,7 +7,8 @@
 		efuse.o		\
 		ps.o		\
 		rc.o		\
-		regd.o
+		regd.o      \
+		stats.o
 
 rtl8192c_common-objs +=		\
 
@@ -24,5 +25,6 @@
 obj-$(CONFIG_RTL8192CU)		+= rtl8192cu/
 obj-$(CONFIG_RTL8192SE)		+= rtl8192se/
 obj-$(CONFIG_RTL8192DE)		+= rtl8192de/
+obj-$(CONFIG_RTL8723AE)		+= rtl8723ae/
 
 ccflags-y += -D__CHECK_ENDIAN__
